File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> char_pos;
        int max_len=0;
        for(int i=0, j=0; j<s.length(); j++){
            if(char_pos.find(s[j])!=char_pos.end() && char_pos[s[j]]>=i){
                i = char_pos[s[j]]+1;
            }
            char_pos[s[j]] = j;
            max_len = max(max_len, j-i+1);
        }
        return max_len;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> charMap = new HashMap<>();
        int start = 0;
        int maxLen = 0;
        for(int end=start; end<s.length(); end++){
            if(charMap.containsKey(s.charAt(end))){
                start = Math.max(start, charMap.get(s.charAt(end))+1);
            }
            charMap.put(s.charAt(end), end);
            maxLen = Math.max(maxLen, end-start+1);
        }
        return maxLen;
    }
}